### ISSUE TITLE
curl does not correctly verify HTTPS server IP against TLS certificate with IP SANs

### ISSUE THREAD

**Issue by omkensey** [coreos/bugs#1513]
# Issue Report #

## Bug ##

curl in the current stable version seems to be overly-strict in its verification of TLS certificates.  I'm running the GCE self-hosted Kubernetes quickstart and when I try to run the init-worker.sh, the script dies with this error:

curl: (51) SSL: no alternative certificate subject name matches target host name '[API server public IP]'

However, dumping the TLS certificate using openssl s_client shows that the IP is in the SAN list.  Running curl manually with the -v option shows this at the critical point:

*  subjectAltName: host "[API server public IP]" matched cert's IP address!
*  subjectAltName does not match [API server public IP]
* SSL: no alternative certificate subject name matches target host name '[API server public IP]'
* Closing connection 0
* TLSv1.2 (OUT), TLS alert, Client hello (1):
curl: (51) SSL: no alternative certificate subject name matches target host name '[API server public IP]'

However, querying the same URL using the same set of TLS assets on a different host (a Google Cloud Shell instance) successfully connects to the API server over TLS with an older version of curl:

```
$ curl -V
curl 7.38.0 (x86_64-pc-linux-gnu) libcurl/7.38.0 OpenSSL/1.0.1t zlib/1.2.8 libidn/1.29 libssh2/1.4.3 librtmp/2.3
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtmp rtsp scp sftp smtp smtps telnet tftp 
Features: AsynchDNS IDN IPv6 Largefile GSS-API SPNEGO NTLM NTLM_WB SSL libz TLS-SRP 
```

### CoreOS Version ###

```
$ cat /etc/os-release
NAME=CoreOS
ID=coreos
VERSION=1068.9.0
VERSION_ID=1068.9.0
BUILD_ID=2016-08-05-0455
PRETTY_NAME="CoreOS 1068.9.0 (MoreOS)"
ANSI_COLOR="1;32"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://github.com/coreos/bugs/issues"
```

### Environment ###

GCE

### Expected Behavior ###

curl properly verifies the TLS cert offered by the server and completes the request.

### Actual Behavior ###

curl fails with error:

curl: (51) SSL: no alternative certificate subject name matches target host name '[API server public IP]'

### Reproduction Steps ###

  1. Using the latest release of bootkube, set up a Kubernetes master on GCE
  2. Run the init-worker.sh script to try to initialize a worker (the TLS assets will be copied to the worker before the script dies).

### Other Information ###

I think this either is, or is related to, https://github.com/curl/curl/issues/875

**Comment by wmsch** [comment]
I see this, too, and it does seem to be a regression from 875.

**Comment by jay** [comment]
Fixed in curl/curl@b6fcdc3.

**Comment by mischief** [comment]
@jay do you expect a release soon to address this..?

**Comment by jay** [comment]
September 7 last I heard. We didn't discuss pushing one out early for this. cc @bagder

**Comment by mischief** [comment]
thanks for the update! :snowman:

**Comment by omkensey** [comment]
Further updates on the curl side in curl/curl#959.

**Comment by omkensey** [comment]
This fix landed in the latest cURL release, [7.50.2](https://curl.haxx.se/changes.html#7_50_2).

**Comment by crawford** [comment]
Fixed via https://github.com/coreos/coreos-overlay/pull/2177.
